### PR TITLE
Improve `astro sync` error handling

### LINKED ISSUE #9711: Malformed entry filename is not reported
### Astro Info

```block
Astro                    v4.1.3
Node                     v18.18.0
System                   Linux (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/sitemap
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When an entry has an invalid frontmatter, the parsing library throws an error but the associated Astro error is irrelevant: it points to the `content/config.*` instead of the faulty file. For example, the following mdx file
```md
---
# example.mdx
title: 'Here's a wrong pair'
---
```
will throw this:
```
[GenerateContentTypesError] `astro sync` command failed to generate content collection types: can not read an implicit mapping pair; a colon is missed
  Hint:
    Check your `src/content/config.*` file for typos.
  Error reference:
    https://docs.astro.build/en/reference/errors/generate-content-types-error/
  Stack trace:
```

### What's the expected result?

Have a new error that specifies what file is failing. I don't think it's realistic to also want the exact location since it depends on the parsing library

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-se3vcm?file=src%2Fcontent%2Fblog%2Ffirst-post.md

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Fixes #9711
- Improves `astro sync` error handling by fixing our `isAstroError()` function. Previously, subclassed errors would not be detected.
- I had a change in mind that I haven't pushed yet and would love to hear thoughts... would it help to centralize all frontmatter parsing into the `@astrojs/internal-helpers` package? `@astrojs/mdx` and `@astrojs/markdoc` currently have their own error handling and it doesn't always match the built-in markdown handling.

## Testing

Tested manually

## Docs

N/A

### CODE REVIEW

**Comment by ematipico** [comment]
What's the reason behind a `minor` change? 🤔

**Review by ematipico — APPROVED** [review]

File: .changeset/healthy-jokes-deny.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"astro": minor
```
Comment (natemoo-re) [inline .changeset/healthy-jokes-deny.md:2]:
```suggestion
"astro": patch
```

**Comment by natemoo-re** [comment]
> What's the reason behind a minor change? 🤔

I think it was just a typo! Changed to `patch`.

**Comment by lilnasy** [comment]
!preview improve-sync-errors

**Review by lilnasy — DISMISSED** [review]
So much better!
image

**Review by lilnasy — APPROVED** [review]

**Pull request merged**